src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs

```diff
@@ -704,5 +707,52 @@ internal ITextBuffer GetOrCreateBufferForPath(string? filePath, IContentType con
                 return textBuffer;
             });
         }
+        internal override bool TryApplyChanges(
+            Solution newSolution,
+            IProgressTracker progressTracker)
+        {
+            var oldSolution = CurrentSolution;
+            if (base.TryApplyChanges(newSolution, progressTracker))
+            {
+                var projectChanges = newSolution.GetChanges(oldSolution).GetProjectChanges().ToList();
+                var changedDocs = projectChanges.SelectMany(pd => pd.GetChangedDocuments(true).Concat(pd.GetChangedAdditionalDocuments())).ToList();
+
+                NotifyRefactorChanges(changedDocs, newSolution, oldSolution);
+                return true;
+            }
+
+            return false;
+        }
+
+        private void NotifyRefactorChanges(
+            IEnumerable<DocumentId> changedDocumentIds,
+            Solution newSolution,
+            Solution oldSolution,
+            CancellationToken cancellationToken = default)
+        {
+            // Without any services to notify there's no need to dig through the documents
+            if (!_refactorNotifyServices.Any())
+            {
+                return;
+            }
+
+            var changedSymbols = Rename.RenameSymbolAnnotation
+                .GatherChangedSymbolsInDocumentsAsync(changedDocumentIds, newSolution, oldSolution, cancellationToken)
+                .WaitAndGetResult_CanCallOnBackground(cancellationToken);
+
+            foreach (var changedSymbolPair in changedSymbols)
+            {
+                var oldSymbol = changedSymbolPair.Key;
+                var newSymbol = changedSymbolPair.Value;
+
+                foreach (var refactorNotifyService in _refactorNotifyServices)
+                {
+                    if (refactorNotifyService.TryOnBeforeGlobalSymbolRenamed(this, changedDocumentIds, oldSymbol, newSymbol.ToDisplayString(), throwOnFailure: false))
+                    {
+                        refactorNotifyService.TryOnAfterGlobalSymbolRenamed(this, changedDocumentIds, oldSymbol, newSymbol.ToDisplayString(), throwOnFailure: false);
```

Comment:
this is so confusing ot me.  why do we have two notifications. why do we call them in succession with nothing in between?

Reply:
shouldn't hte `TryOnBefore` be called *prior* to us calling base.TryApplyChanges, then the `TryOnAfter` is called after that base call returns?

Reply:
It looks like there were a few instances in code where it was like this, but I believe calling `TryOnBefore` > `ApplyChanges` > `TryOnAfter` is the correct way to do this